x/cosmwasmpool/gov.go

```diff
@@ -0,0 +1,137 @@
+		}
+	}
+}
+
+// uploadCodeIdAndWhitelist uploads the given wasm bytecode to the wasmvm. Whitelists the resulting code id
+// Emits an event with the code id and checksum.
+// Returns error if byte code is empty or if fails to upload the code.
+func (k Keeper) uploadCodeIdAndWhitelist(ctx sdk.Context, byteCode []byte) (uint64, error) {
+	if len(byteCode) == 0 {
+		return 0, fmt.Errorf("empty wasm bytecode")
+	}
+
+	cosmwasmPoolModuleAddress := k.accountKeeper.GetModuleAddress(types.ModuleName)
+
+	// Only allow the x/cosmwasmpool module to instantiate this contract.
+	instantiatePermissions := wasmtypes.AccessConfig{
+		Permission: wasmtypes.AccessTypeAnyOfAddresses,
+		Addresses:  []string{cosmwasmPoolModuleAddress.String()},
+	}
+
+	// Upload the code to the wasmvm.
+	codeID, checksum, err := k.contractKeeper.Create(ctx, cosmwasmPoolModuleAddress, byteCode, &instantiatePermissions)
+	if err != nil {
+		return 0, err
+	}
+
+	// Add the code id to the whitelist.
+	k.WhitelistCodeId(ctx, codeID)
+
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.TypeEvtUploadedCosmwasmPoolCode,
+		sdk.NewAttribute(types.AttributeKeyCodeID, strconv.FormatUint(codeID, 10)),
+		sdk.NewAttribute(types.AttributeKeyChecksum, string(checksum)),
+	))
+
+	return codeID, nil
+}
+
+// migrateComswasmPools migrates all given cw pool contracts specified by their IDs.
+// It has two options to perform the migration.
+//
+// 1. If the codeID is non-zero, it will migrate the pool contracts to a given codeID assuming that it has already
+// been uploaded. uploadByteCode must be empty in such a case. Fails if codeID does not exist.
+// Fails if uploadByteCode is not empty.
+//
+// 2. If the codeID is zero, it will upload the given uploadByteCode and use the new resulting code id to migrate
+// the pool to. Errors if uploadByteCode is empty or invalid.
+//
+// In both cases, if one of the pools specified by the given poolID does not exist, the proposal fails.
+//
+// The reason for having poolIDs be a slice of ids is to account for the potential need for emergency migration
+// of all old code ids associated with particular pools to new code ids, or simply having the flexibility of
+// migrating multiple older pool contracts to a new one at once when there is a release.
+//
+// poolD count to be submitted at once is gated by a governance paramets (20 at launch).
+// The proposal fails if more. Note that 20 was chosen arbitrarily to have a constant bound on the number of pools migrated
+// at once. This size will be configured by a module parameter so it can be changed by a constant.
+func (k Keeper) migrateCosmwasmPools(ctx sdk.Context, poolIds []uint64, newCodeId uint64, uploadByteCode []byte, migrateMsg []byte) (err error) {
+	cosmwasmPoolModuleAddress := k.accountKeeper.GetModuleAddress(types.ModuleName)
+
+	if err := types.ValidateMigrationProposalConfiguration(poolIds, newCodeId, uploadByteCode); err != nil {
```

Comment:
after this point we know we either have a code id or a bytecode, but not both. Maybe we should add a comment saying that.